### PR TITLE
rtl_tcp: favorites slide-out panel (closes #315)

### LINKED ISSUE #315: rtl_tcp UI: move favorites to a dedicated slide-out panel
Part of epic #299. Refines the client-picker UI in **#303** (Linux GTK) and should inform the corresponding mac SwiftUI work in **#304**.

## Current plan (baseline in #303)

The client picker in #303 puts favorites inline in the source-panel sidebar alongside the auto-discovered list and manual entry:

> - **Favorites / known servers**: remembered across app restarts; star icon to pin a discovered server or manual entry

Favorites sit in the sidebar always-visible, competing with tuning controls for vertical space.

## Proposed refinement

Move the favorites list out of the sidebar and into a dedicated slide-out panel, reachable two ways:

1. **Icon in the header bar** (star / bookmark-shape) that toggles the slide-out
2. **"Manage favorites…" link** from the source-panel itself

When the slide-out is open, it takes over the controls region (or overlays it via `AdwOverlaySplitView` stacking / a `GtkPopover`). Closes by clicking the icon again, pressing Esc, or clicking outside.

## Why

- **Reclaims sidebar real estate** for tuning controls (gain, PPM, bias-T, direct sampling, etc. — the set enumerated in #303). Those controls are the dense area; favorites is not.
- **Richer favorite entries** become affordable: per-server nickname, tuner type, last-seen timestamp, bandwidth warning indicator, recent-tuning preview, "copy share URL" action. Hard to fit in a cramped sidebar row.
- **Matches GNOME UX patterns**: bookmarks-style sidebars are familiar (Files, Epiphany, Fractal).

## Tradeoff

One extra click to connect. For a first-time user that's fine (discover → try → star). For a daily user who opens the app to hop on "home scanner server," the slide-out adds friction.

Two mitigations worth deciding on during impl:

- **(a) Auto-open the slide-out on first launch** if no source is active — makes the "I opened the app, where do I click?" moment obvious without forcing the slide-out on later launches.
- **(b) "Last connected" quick button in the main source panel** — a single one-click jump-to-home entry, with the full favorites list behind the slide-out icon. Covers the 90% case without the extra click.

Recommendation: ship both (a) and (b). Cheap, orthogonal, and they address the two distinct user states (cold start vs. returning).

## Scope interaction with #303

When #303 is implemented, its "favorites inline" section is replaced by this slide-out pattern. The state store (pinned entries, last-connected, nicknames) stays identical; only the presentation layer changes. If this issue lands alongside #303, merge the two designs in the implementation. If it lands later, #303 ships with inline favorites and this refactors the UI.

## Mac side (#304)

SwiftUI analog: `NavigationSplitView` already supports a trailing inspector/sidebar, or a toolbar `Button` toggling a `Sheet` for the slide-out. Same information architecture; different API. Worth mirroring so Linux and mac users share the same mental model.

## Not in scope

- Importing/exporting favorites (future enhancement, out of epic #299)
- Categorizing favorites into folders (probably YAGNI for SDR use)

### PR DESCRIPTION
**Pull request by jasonherald**
Implements #315 — moves rtl_tcp favorites out of the inline sidebar star skeleton and into a dedicated header-bar slide-out popover. Richer per-entry metadata, two entry points, full per-row actions.

## Summary

- **Schema**: favorites now persist as a `FavoriteEntry { key, nickname, tuner_name, gain_count, last_seen_unix }` list instead of bare strings. Loader accepts both shapes for backward compat (upgrades legacy entries to stubs on read; skips corrupt object entries with a warn log).
- **Header-bar button**: star-icon `GtkMenuButton` with attached `GtkPopover` containing a scrollable `GtkListBox`. Empty-state label shown when no favorites are pinned.
- **Popover rows**: each `FavoriteEntry` renders as an `AdwActionRow` with nickname title and `host:port • tuner · N gains • seen Xm ago` subtitle. Three suffix buttons: **Connect** (pins TCP before host/port writes per the #335 ordering rule, dispatches `SetNetworkConfig` + conditional `SetSourceType`, persists `LastConnectedServer`, dismisses popover), **Copy** (`host:port` → display clipboard), **Unstar** (removes from map, persists, reorders discovery expander, rebuilds popover).
- **Second entry point**: "Manage favorites…" button packed as a flat suffix on the RTL-TCP `Discovered rtl_tcp servers` expander. Click calls `MenuButton::popup()` so the popover appears anchored to the header regardless of which button fired the click.
- **Widget-closure hygiene**: all per-row action deps bundle into `FavoriteRowContext` — each button closure clones a single `Rc` instead of capturing nine weak refs. Popover widgets held via `glib::WeakRef` (the `ServerStatusWidgetsWeak` / #329 pattern) so the closures stay drop-safe.

## Commits

1. `feat(#315): extend favorite schema with rich entry data` — `FavoriteEntry` struct, dual-shape loader, round-trip + upgrade + corrupt-object tests
2. `feat(#315): header-bar favorites button + empty popover scaffolding` — `FavoritesHeaderHandle` + `build_favorites_header`
3. `feat(#315): populate favorites popover from the in-memory map` — `FavoritesPopoverWeak` + `rebuild` method + seed-on-startup + refresh hooks in star-toggle and re-announce paths
4. `feat(#315): per-row Connect / Copy / Unstar actions on favorites popover` — `FavoriteRowContext`, free-fn rebuild, `attach_favorite_row_actions`, `parse_host_port` + its 6 unit tests
5. `feat(#315): Manage favorites… button in RTL-TCP source panel` — second entry point wired to `MenuButton::popup()`

## Test plan

- [x] `cargo build --workspace` clean
- [x] `cargo clippy --all-targets --workspace -- -D warnings` clean
- [x] `cargo test --workspace` — all tests pass, including 10 new tests (favorites schema round-trip / legacy upgrade / corrupt-object skip / `now_unix_seconds` monotonicity / 6x `parse_host_port` cases)
- [x] `cargo fmt -- --check` clean
- [x] `cargo deny` + `cargo audit` clean
- [x] Smoke tested on Linux + `sherpa-cuda`: header star opens popover; empty-state label shown with no favorites; star from discovery row adds to popover; Connect button tunes correctly and dismisses popover; Copy button writes `host:port` to clipboard; Unstar button removes from list and updates discovery ordering; "Manage favorites…" button in source panel opens same popover; favorites persist across app restart with full `FavoriteEntry` metadata
- [x] Discovery-list regression check: mDNS rows still appear, Connect from discovery row still works, star-toggle from discovery row still refreshes popover

## Known transient UI behavior

When a favorite is unstarred from the popover while the corresponding discovery row is visible, that row's star-button icon is NOT refreshed immediately — it self-heals on the next mDNS re-announce (which rebuilds the row with `set_active(is_favorite)`). Documented inline; acceptable edge case since the simultaneously-visible popover-unstar + discovery-list view is uncommon.

## Out of scope / follow-ups

- Frequency bookmarks slide-out — filed as **#339** (right-side slide-out mirroring this pattern, plus filter/search row and `rr_category` grouping in v1)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Header-bar favorites popover with scrollable list and per-row Connect/Copy/Unstar actions; added a "Manage Favorites" button in the sidebar and a second entry in discovery rows.
  * Favorites show rich metadata (nickname, tuner, gain, last-seen) and refresh when servers re-announce; favorites sort deterministically.

* **Bug Fixes**
  * More reliable connect sequencing and UI refresh on changes; improved persistence handling.

* **Tests**
  * Unit tests covering favorites persistence, legacy upgrades, corrupt-entry handling, parsing, and time utilities.

### CODE REVIEW

File: crates/sdr-ui/src/sidebar/source_panel.rs [hunk 1]
```diff
@@ -543,30 +565,113 @@ pub struct LastConnectedServer {
...
+                    Some(FavoriteEntry {
+                        key: s.to_string(),
+                        nickname: s.to_string(),
+                        tuner_name: None,
+                        gain_count: None,
+                        last_seen_unix: None,
+                    })
+                } else {
+                    serde_json::from_value::<FavoriteEntry>(entry.clone()).ok()
+                }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/sidebar/source_panel.rs]: Fixed in 5aef45a. The schema docstring promised a warn log but the code silently used `.ok()` — thanks for catching the drift. Switched to a `match` with `tracing::warn!` that logs the entry index and the serde error, still returns `None` so the rest of the list loads. Existing `favorites_loader_skips_corrupt_object_entries` test continues to pass (it only asserts on the surviving entry, not on the log side-effect).

File: crates/sdr-ui/src/window.rs [hunk 2]
```diff
@@ -1224,15 +1400,28 @@ fn connect_rtl_tcp_discovery(
...
+                        server.txt.nickname.clone()
+                    };
+                    let star_tuner_name = Some(server.txt.tuner.clone());
+                    let star_gain_count = Some(server.txt.gains);
                     let star_favorites = Rc::clone(&favorites);
                     let star_config = std::sync::Arc::clone(&config_for_discovery);
                     let star_rows = Rc::clone(&displayed_rows);
                     let star_expander_weak = expander_weak.clone();
+                    let star_row_ctx = Rc::clone(&favorite_row_ctx);
                     star_btn.connect_toggled(move |btn| {
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs]: Fixed in 5aef45a. Confirmed the cycle: the HashMap holds strong GObject refs on each `AdwActionRow`, every row's `connect_toggled` closure captured `Rc<RefCell<HashMap<...>>>` both directly (`star_rows`) and transitively via `FavoriteRowContext.displayed_rows`, closing the loop map → row → closure → ctx → map. On window teardown, nothing dropped.

Changed `FavoriteRowContext.displayed_rows` to `std::rc::Weak`, dropped the redundant `star_rows` clone on the star_btn closure, and reach the map via `star_row_ctx.displayed_rows.upgrade()` at use time — same pattern in the unstar closure. The discovery-polling `glib::timeout_add_local` timer retains the strong `Rc`, so upgrades succeed while the timer is running and correctly return `None` after teardown (state in which there's nothing to reorder). Added a docstring on the `displayed_rows` field to record the retain-cycle reasoning for future readers.

File: crates/sdr-ui/src/sidebar/source_panel.rs [hunk 3]
```diff
@@ -543,30 +565,129 @@ pub struct LastConnectedServer {
...
+/// Load the persisted favorites list. Returns an empty `Vec` on
+/// first launch / absent / corrupt config — safe to call
+/// unconditionally.
+///
+/// **Backward compatibility:** accepts two on-disk shapes:
+///
+/// 1. **Current (PR #315):** `Vec<FavoriteEntry>` — array of JSON
+///    objects, each decoded via `serde_json::from_value`. Objects
+///    that fail to deserialize are silently skipped (same
+///    schema-drift tolerance as `load_last_connected`).
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/sidebar/source_panel.rs]: Fixed in 4d9846a. Updated the doc bullet to state that failed `serde_json::from_value` entries are skipped AND logged at `tracing::warn!` with entry index + serde error — matches the actual behavior so schema drift stays diagnosable from the API docs without reading the impl.

File: crates/sdr-ui/src/sidebar/source_panel.rs [hunk 4]
```diff
@@ -697,31 +818,107 @@ mod tests {
...
+        assert_eq!(loaded[0].key, "shack-pi.local.:1234");
+    }
+
+    #[test]
+    fn now_unix_seconds_is_monotonic_within_call() {
+        // Not a real monotonicity test — just a smoke-test that
+        // the helper returns a sensible modern value. Anything
+        // past 2020-01-01 (Unix 1_577_836_800) is clearly real
+        // wall-clock time and not a clock-skew fallback.
+        assert!(now_unix_seconds() > 1_577_836_800);
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/sidebar/source_panel.rs]: Fixed in 4d9846a. Hoisted the two load-bearing literals to named consts at the top of the test module:

```rust
const TEST_LAST_SEEN_UNIX: u64 = 1_700_000_000;
const MODERN_UNIX_FLOOR: u64 = 1_577_836_800;  // 2020-01-01T00:00:00Z
```

The round-trip test now reads `Some(TEST_LAST_SEEN_UNIX)` and the monotonicity smoke test reads `> MODERN_UNIX_FLOOR`. Intent is self-documenting and matches the repo's named-constants rule.

File: crates/sdr-ui/src/window.rs [hunk 5]
```diff
@@ -1047,17 +1186,56 @@ fn connect_rtl_tcp_discovery(
...
+        device_row: device_row.downgrade(),
+        expander_weak: expander_weak.clone(),
+        // Weak ref — see `FavoriteRowContext.displayed_rows`
+        // docstring for the retain-cycle reasoning.
+        displayed_rows: Rc::downgrade(&displayed_rows),
+    });
+    // Seed the popover's content from the restored favorites so
+    // the list is ready when the user first clicks the header
+    // star, without waiting for a mutation to trigger a rebuild.
+    rebuild_favorites_popover(&favorite_row_ctx, &favorites.borrow());
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs:1262]: Fixed in 4d9846a. Real bug — rebuild-on-mutation misses the quiescent case where the user closes the popover and the subtitle ages go stale until the next mDNS beacon.

Added a `popover.connect_show` handler that rebuilds the list on every popup, so `format_favorite_subtitle` re-snapshots `now_unix` against current wall-clock. No retain cycle: the closure captures `Rc<FavoriteRowContext>` whose `popover` field is already weak (and the map is weak per round 1), so `popover → signal → closure → ctx → { weak popover, weak map }` terminates rather than looping back.

File: crates/sdr-ui/src/sidebar/source_panel.rs [hunk 6]
```diff
@@ -543,30 +565,131 @@ pub struct LastConnectedServer {
...
+/// `window.rs::favorite_key`) alongside display metadata the
+/// favorites slide-out shows even when the server is offline: the
+/// nickname the user last saw, the tuner type and gain-step count
+/// from the last mDNS announcement, and a "last seen" wall-clock
+/// stamp.
+///
+/// Optional fields default to `None` so a freshly-starred server
+/// with no cached metadata still round-trips correctly, and so
+/// legacy bare-string entries (PR #335 schema) can be read back
+/// without drift — see `load_favorites`.
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/sidebar/source_panel.rs:587]: Fixed in e2b52bd. Rewrote the `KEY_RTL_TCP_CLIENT_FAVORITES` docstring to describe the current schema: JSON array of [`FavoriteEntry`] objects keyed by the stable `hostname:port` identity from `window.rs::favorite_key`. Spelled out why we key by host:port and NOT `instance_name` (which is operator-editable and would silently drop the star on any rename — your learning reference is exactly that invariant). Also noted both the legacy bare-string fallback and the corrupt-entry `tracing::warn!` on the read path, so the contract is fully discoverable from the constant's docs.

File: crates/sdr-ui/src/window.rs [hunk 7]
```diff
@@ -1448,6 +1731,320 @@ fn reorder_discovered_rows(
...
+    let secs = now_unix - last_seen_unix;
+    if secs < 60 {
+        "just now".to_string()
+    } else if secs < 3_600 {
+        format!("{}m ago", secs / 60)
+    } else if secs < 86_400 {
+        format!("{}h ago", secs / 3_600)
+    } else {
+        format!("{}d ago", secs / 86_400)
+    }
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs:2191]: Fixed in e2b52bd. Hoisted the three boundaries into named constants above `format_seen_age`, using a derived-value formulation so the relationships are obvious:

```rust
const SECONDS_PER_MINUTE: u64 = 60;
const SECONDS_PER_HOUR: u64 = 60 * SECONDS_PER_MINUTE;
const SECONDS_PER_DAY: u64 = 24 * SECONDS_PER_HOUR;
```

Kept in raw `u64` Unix-seconds (not `std::time::Duration`) because `last_seen_unix` lives in the favorites JSON as `u64` seconds end-to-end; a Duration-based variant would round-trip through `as_secs()` at every site. Docstring on the const block calls that out.

File: crates/sdr-ui/src/window.rs [hunk 8]
```diff
@@ -1448,6 +1731,328 @@ fn reorder_discovered_rows(
...
+    if secs < SECONDS_PER_MINUTE {
+        "just now".to_string()
+    } else if secs < SECONDS_PER_HOUR {
+        format!("{}m ago", secs / SECONDS_PER_MINUTE)
+    } else if secs < SECONDS_PER_DAY {
+        format!("{}h ago", secs / SECONDS_PER_HOUR)
+    } else {
+        format!("{}d ago", secs / SECONDS_PER_DAY)
+    }
+}
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs:2192]: Fixed in b23b278. Added a new `favorite_subtitle_format_tests` module alongside the existing `rtl_tcp_discovery_format_tests` with 10 tests pinning:

- `format_seen_age`: sub-minute "just now", integer-division minute bucket (179s → 2m, not 3m), hour/day boundary cases (3_599 vs 3_600, 86_399 vs 86_400), and the clock-skew branch (`last_seen >= now` → "just now", no underflow)
- `format_favorite_subtitle`: full metadata path, missing-tuner and missing-gains both drop the middle segment (the `if let (Some, Some)` both-or-nothing rule), `last_seen_unix` None vs Some(0) both render as "offline" (the `ts > 0` guard stops a default-valued timestamp from rendering as "seen 55y ago" against the 1970 epoch)

Determinism via a fixed `NOW_UNIX = 1_700_000_000` const — all offsets derive from there.

**Comment by jasonherald** [comment]
@coderabbitai Re: [the outside-diff-range "Extract the shared RTL-TCP connect sequence" comment](https://github.com/jasonherald/rtl-sdr/pull/340#pullrequestreview-4136233826) — fixed in b23b278 alongside the subtitle tests.

Extracted `apply_rtl_tcp_connect` that owns the ordering-sensitive flow (snapshot `already_rtl_tcp`, pin TCP on `protocol_row` BEFORE host/port writes, flip to RTL-TCP, dispatch `SetNetworkConfig`, persist `LastConnectedServer`, conditional `SetSourceType(RtlTcp)`). Both the discovery-row Connect closure and the favorites-popover Connect closure now call it — the helper's doc spells out the load-bearing reasoning for each step so the next ordering fix lands in one place.

The favorites popover's popover `popdown()` stays as a caller-owned follow-up after the helper returns, since it's popover-specific.

File: crates/sdr-ui/src/window.rs [hunk 9]
```diff
@@ -870,9 +904,91 @@ fn build_header_bar(
...
+    content.append(&title);
+    content.append(&empty_label);
+    content.append(&scroll);
+    popover.set_child(Some(&content));
+
+    let button = gtk4::MenuButton::builder()
+        .icon_name("starred-symbolic")
+        .tooltip_text("Pinned rtl_tcp servers")
+        .popover(&popover)
+        .build();
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs:982]: Fixed in 58d98c0. Added explicit `GtkAccessible` `Label` property via `widget.update_property(&[gtk4::accessible::Property::Label(...)])` on all four icon-only controls:

- Header MenuButton (starred-symbolic): `"Pinned servers menu"`
- Popover Copy button (edit-copy-symbolic): `"Copy server address"`
- Popover Unstar button (starred-symbolic): `"Remove from favorites"`
- Discovery-row pin toggle: **state-dependent** — `"Pin as favorite"` when unpinned, `"Unpin from favorites"` when pinned. Extracted into `set_favorite_toggle_accessible_name(btn, is_favorite)` and called from two sites: row-build (off `starred_initially`) and the toggled closure (off fresh `is_active()`). The label describes the NEXT action, not the icon's current appearance, which is what AT users want to hear before clicking.

File: crates/sdr-ui/src/window.rs [hunk 10]
```diff
@@ -1448,6 +1768,311 @@ fn reorder_discovered_rows(
...
+    list.set_visible(has_any);
+    if !has_any {
+        return;
+    }
+    let now = sidebar::source_panel::now_unix_seconds();
+    let mut entries: Vec<&sidebar::source_panel::FavoriteEntry> = favorites.values().collect();
+    // Alpha by nickname so adding / removing doesn't reorder
+    // the rest of the list.
+    entries.sort_by_key(|e| e.nickname.to_lowercase());
+    for entry in entries {
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs:1957]: Fixed in 58d98c0. Primary key still the lowercased nickname; added `.then_with(|| a.key.cmp(&b.key))` so duplicate-nickname entries fall back to the stable hostname:port key — cleans up the HashMap-iteration-order reshuffle across inserts / restarts.

Extracted the sort into a pure `sort_favorites_for_display` free fn (testable without GTK widgets) and added 3 unit tests: lowercased-nickname primary order, tie-break-on-key when three entries share a nickname, idempotent-when-already-sorted smoke test.

File: crates/sdr-ui/src/window.rs [hunk 11]
```diff
@@ -762,11 +770,29 @@ fn build_sidebar_toggle(split_view: &adw::OverlaySplitView) -> gtk4::ToggleButto
...
+/// `ListBox` inside that popover — `connect_rtl_tcp_discovery`
+/// clears + re-populates it when the favorites map changes. The
+/// `empty_label` is shown when the list is empty so the user sees
+/// "No pinned servers yet" instead of a blank popover.
+pub struct FavoritesHeaderHandle {
+    pub button: gtk4::MenuButton,
+    pub popover: gtk4::Popover,
+    pub list: gtk4::ListBox,
+    pub empty_label: gtk4::Label,
+}
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs]: Fixed in 47bfe51. Confirmed no external module touches the struct or its fields (grepped for `FavoritesHeaderHandle` and `favorites_handle.*` / `favorites_header.*` — all hits inside `window.rs`), so dropped the `pub` on both the struct and all four fields. Keeps the header/popover layout as a strictly internal detail.

File: crates/sdr-ui/src/window.rs [hunk 12]
```diff
@@ -1448,6 +1818,316 @@ fn reorder_discovered_rows(
...
+            unstar_ctx.displayed_rows.upgrade(),
+        ) {
+            reorder_discovered_rows(&expander, &rows.borrow(), &unstar_ctx.favorites.borrow());
+        }
+        // Rebuild the popover so the unstarred row disappears.
+        // GTK signal-lifetime guarantees we can `ListBox::remove`
+        // our own row from inside this button-clicked handler:
+        // GTK retains the signal's source widget for the
+        // callback's duration, so the button won't drop under us.
+        rebuild_favorites_popover(&unstar_ctx, &unstar_ctx.favorites.borrow());
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/sdr-ui/src/window.rs:2125]: Good catch — you're right this wasn't just cosmetic. Fixed in 47bfe51.

Added an auxiliary `discovered_star_buttons` map keyed by `favorite_key` (hostname:port) holding `glib::WeakRef<ToggleButton>`, populated when each discovery row builds its star. `FavoriteRowContext` gains a `std::rc::Weak` ref on the map (same retain-cycle rationale as the existing `displayed_rows` weak — the ToggleButton's closure captures the context, and a strong back-ref would loop).

The unstar closure now upgrades and looks up by key: if a live discovery row exists and its toggle is currently active, `set_active(false)` — which fires the toggle's own `connect_toggled` handler, which runs the idempotent remove / persist / reorder / popover rebuild. Unstar early-returns there so we don't duplicate that work. When no discovery row is visible (withdrawn server) the fallback path still does the reorder + rebuild directly.

Self-cleaning: stale entries in the star-button map from dropped rows fail to upgrade silently and overwrite on the next row re-announce.
